src/java/org/apache/cassandra/service/accord/serializers/CommandStoreSerializers.java

```diff
@@ -153,9 +153,9 @@ public RedundantBefore.Entry deserialize(DataInputPlus in, int version) throws I
             long endEpoch = in.readUnsignedVInt();
             if (endEpoch == 0) endEpoch = Long.MAX_VALUE;
             else endEpoch = startEpoch + 1 + endEpoch;
-            TxnId bootstrappedAt = CommandSerializers.txnId.deserialize(in, version);
             TxnId locallyAppliedOrInvalidatedBefore = CommandSerializers.txnId.deserialize(in, version);
             TxnId shardAppliedOrInvalidatedBefore = CommandSerializers.txnId.deserialize(in, version);
+            TxnId bootstrappedAt = CommandSerializers.txnId.deserialize(in, version);
```

Comment:
do we have any test that now detects this?  If not we should add one

Reply:
This was found by new invariants. Happy to add a dedicated test if you think that would be warranted as well, but will not have time to do a comprehensive end to end test of serializers to catch this kind of problem (but maybe should take an action item to do so comprehensively before this is considered prod ready)

Reply:
I am working on a test, and seems that there is still an issue...

the equality check failed here:

```
expected: {(tbl:-Inf,tbl:pk1]=(195970445814877,195970445814878,[195970445814808,7046332644952687526,9(RX),1277444686]), (tbl:pk1,tbl2:pk2]=null, (tbl2:pk2,tbl2:pk3]=(228431633815866,228431633815867,...
 but was: {(tbl:-Inf,tbl:pk1]=(195970445814877,195970445814880,[195970445814808,7046332644952687526,9(RX),1277444686]), (tbl:pk1,tbl2:pk2]=null, (tbl2:pk2,tbl2:pk3]=(228431633815866,228431633815869,...
```

Reply:
updated the reflective diff to work with arrays and see

```
ReflectionUtils.recursiveEquals(read, input);
diff = {ArrayList@3937}  size = 7
 0 = {ReflectionUtils$Difference@4754} "Diff{path='.values.[0].endEpoch.', lhs=195970445814880, rhs=195970445814878}"
 1 = {ReflectionUtils$Difference@4755} "Diff{path='.values.[2].endEpoch.', lhs=228431633815869, rhs=228431633815867}"
 2 = {ReflectionUtils$Difference@4756} "Diff{path='.values.[4].endEpoch.', lhs=278679697445769, rhs=278679697445767}"
 3 = {ReflectionUtils$Difference@4757} "Diff{path='.values.[6].endEpoch.', lhs=274102584243720, rhs=274102584243718}"
 4 = {ReflectionUtils$Difference@4758} "Diff{path='.values.[8].endEpoch.', lhs=263275610724048, rhs=263275610724046}"
 5 = {ReflectionUtils$Difference@4759} "Diff{path='.values.[10].endEpoch.', lhs=201513512660742, rhs=201513512660740}"
 6 = {ReflectionUtils$Difference@4760} "Diff{path='.values.[12].endEpoch.', lhs=277526568649263, rhs=277526568649261}"
```

so for some reason every other entry has the wrong epoch

Reply:
Sent out a feedback branch with the bug fixes

Accord: https://github.com/belliottsmith/cassandra-accord/pull/9
C*: https://github.com/belliottsmith/cassandra/pull/6

Reply:
Great catch, will take a look shortly.

Reply:
Merged, rebased and force pushed. Thanks!